File: services/fare-pricing/src/fare_pricing/ports/messaging.py
from __future__ import annotations

from typing import Protocol

from . import EventEnvelope


class PublishFailed(Exception):
    """Raised when the event bus could not accept a published event."""


class EventPublisher(Protocol):
    """Abstract port for publishing domain events to the event bus.

    Implementation rules (see docs/08-contracts/messaging.md):
    - Target stream is derived from envelope.producer -> "events:<producer>".
    - Serialize the entire envelope as a single JSON value in the "envelope" field.
    - Apply retention policy (MAXLEN ~ 100000) on publish.
    - Retry with exponential backoff (3 attempts) on transient failure.
    - Must be thread-safe.
    """

    def publish(self, envelope: EventEnvelope) -> None:
        """Publish an event. Raises PublishFailed on persistent failure."""
        ...


class SubscribeFailed(Exception):
    """Raised when the subscriber could not start."""


class HandlerError(Exception):
    """Base for errors returned by the event handler."""


class TransientHandlerError(HandlerError):
    """Transient error; the message should be retried."""


class FatalHandlerError(HandlerError):
    """Fatal error; the message should be moved to DLQ."""


class EventSubscriber(Protocol):
    """Abstract port for subscribing to event streams as a consumer group member.

    Implementation rules (see docs/08-contracts/messaging.md):
    - Create consumer group on startup (XGROUP CREATE ... MKSTREAM).
    - Poll with XREADGROUP in a loop.
    - Call handler with deserialized EventEnvelope.
    - XACK on success, do NOT XACK on transient error, move to DLQ on fatal.
    - Track delivery attempts; move to DLQ after 5 attempts.
    - Run XAUTOCLAIM every 60s for recovery.
    - Graceful shutdown support.
    """

    def subscribe(
        self,
        streams: list[str],
        group: str,
        consumer_name: str,
        handler: EventHandler,
    ) -> None:
        """Start subscribing. Raises SubscribeFailed on startup failure."""
        ...


class EventHandler(Protocol):
    """Callback invoked by EventSubscriber for each received event."""

    def __call__(self, envelope: EventEnvelope) -> None:
        """Process an event. Raises TransientHandlerError or FatalHandlerError."""
        ...

File: services/offer-management/src/ports/messaging.ts
// ---------------------------------------------------------------------------
// EventEnvelope — shared envelope for all domain events
// Per shared-primitives.md §1 and messaging.md Abstract Ports
// ---------------------------------------------------------------------------

export type EventEnvelope = Readonly<{
  eventId: string;
  eventType: string;
  schemaVersion: number;
  producer: string;
  causationId?: string;
  correlationId: string;
  occurredAt: string;
  payload: Record<string, unknown>;
}>;

// ---------------------------------------------------------------------------
// EventPublisher port — application layer depends ONLY on this interface
// ---------------------------------------------------------------------------

export class PublishFailed extends Error {
  constructor(message: string, public readonly cause?: unknown) {
    super(message);
    this.name = "PublishFailed";
  }
}

export interface EventPublisher {
  /**
   * Publish a domain event to the event bus.
   *
   * - The implementation MUST determine the target stream from the `producer`
   *   field of the envelope (stream = "events:<producer>").
   * - The implementation MUST serialize the entire envelope as a single JSON
   *   value in the "envelope" field of the Redis Stream entry.
   * - The implementation MUST apply the retention policy (MAXLEN ~ 100000).
   * - On transient failure, retry with exponential backoff (3 attempts).
   * - On persistent failure, throw PublishFailed.
   */
  publish(envelope: EventEnvelope): Promise<void>;
}

// ---------------------------------------------------------------------------
// EventSubscriber port — application layer depends ONLY on this interface
// ---------------------------------------------------------------------------

export class SubscribeFailed extends Error {
  constructor(message: string, public readonly cause?: unknown) {
    super(message);
    this.name = "SubscribeFailed";
  }
}

export type HandlerResult = "ack" | "retry" | "dlq";

/**
 * Event handler callback — returns a result indicating how the adapter
 * should handle the message (ack, retry, or move to DLQ).
 */
export type EventHandler = (envelope: EventEnvelope) => Promise<HandlerResult>;

export interface EventSubscriber {
  /**
   * Subscribe to one or more event streams as a consumer group member.
   *
   * The implementation runs in the background and calls the handler for each
   * received message. The handler returns ack/retry/dlq to guide ack behaviour.
   *
   * Lifecycle: the subscriber runs until the provided AbortSignal is triggered.
   */
  subscribe(
    streams: readonly string[],
    group: string,
    consumerName: string,
    handler: EventHandler,
    signal: AbortSignal,
  ): Promise<void>;
}
